### PR TITLE
Add initial implementation of color cycling on tiles

### PR DESCRIPTION
**Pull request by ihhub**
relates to #289

This is a initial solution towards cycling of colors on World Map.

### CODE REVIEW

File: src/fheroes2/agg/pal.cpp [hunk 1]
```diff
@@ -162,6 +162,22 @@ namespace PAL
     }
 }
 
+std::vector<uint8_t> PAL::GetCyclingPalette( int stepId )
+{
+    std::vector<uint8_t> palette = PAL::GetPalette( PAL::STANDARD );
```
Comment (ihhub) [inline src/fheroes2/agg/pal.cpp:167]: Hi @idshibanov , could you please take a look how animation is done for water surface. Looks like we are missing something in color cycling as water in the original game is animated slightly differently?

File: src/engine/surface.cpp [hunk 2]
```diff
@@ -608,6 +608,13 @@ void Surface::SetPalette( void )
     }
 }
 
+void Surface::SetPalette( std::vector<SDL_Color> & palette )
+{
+    if ( isValid() && !palette.empty() && surface->format->palette ) {
```
Comment (ihhub) [inline src/engine/surface.cpp:613]: We might remove this function in the future and write a function to draw 8-bit images onto 32-bit images directly.

File: src/fheroes2/agg/pal.cpp [hunk 3]
```diff
@@ -290,3 +306,16 @@ void PAL::Clear()
...
     standard_palette.clear();
 }
+
+std::vector<SDL_Color> PAL::GetCustomSDLPalette( const std::vector<uint8_t> & indexes )
+{
+    if ( indexes.size() != PALETTE_SIZE )
+        return std::vector<SDL_Color>();
+
+    std::vector<SDL_Color> sdlPalette( PALETTE_SIZE );
+    const std::vector<SDL_Color> & colors = *current_palette->colors;
```
Comment (idshibanov) [inline src/fheroes2/agg/pal.cpp:316]: @ihhub the bug is on this line here - you should use base palette, not the current one.

File: src/fheroes2/agg/pal.cpp (hunk 1 above)
Comment (idshibanov) [reply src/fheroes2/agg/pal.cpp:167]: Output of this function is fine, exactly what we need.

File: src/engine/surface.cpp (hunk 2 above)
Comment (idshibanov) [reply src/engine/surface.cpp:613]: @ihhub Looks like `SDL_SetPaletteColors` call overwrites `standard_palette` vector. So this function is part of the problem - if we can get rid of it then perfect. Another solution is to convert `current_palette` pointer into a vector copy.

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 4]
```diff
@@ -163,6 +168,14 @@ void Interface::GameArea::Redraw( Surface & dst, int flag ) const
 
 void Interface::GameArea::Redraw( Surface & dst, int flag, const Rect & rt ) const
 {
+    if ( Game::AnimateInfrequentDelay( Game::COLOR_CYCLE_BATTLE_DELAY ) ) {
+        const std::vector<SDL_Color> tempPalette = PAL::GetCustomSDLPalette( PAL::GetCyclingPalette( Game::MapsAnimationFrame() ) );
+        if ( tempPalette.size() == _sdlPalette.size() ) {
+            std::memcpy( _sdlPalette.data(), tempPalette.data(), sizeof( SDL_Color ) * PALETTE_SIZE );
+            _updatePalette = true;
+        }
+    }
```
Comment (idshibanov) [inline src/fheroes2/gui/interface_gamearea.cpp:177]: Also I think you should move this code out of `Redraw` and into a separate function that will be called from `game.cpp` when `maps_animation_frame` is updated.

You don't need an additional COLOR_CYCLE_BATTLE_DELAY check.

Comment (ihhub) [reply src/fheroes2/gui/interface_gamearea.cpp:177]: We have the check as Redraw could be called during scrolling causing an increased animation ;)

**Comment by idshibanov** [comment]
@ihhub I can make required changes here If you want, already debugged it.

**Comment by ihhub** [comment]
> @ihhub I can make required changes here If you want, already debugged it.

Hi @idshibanov , yes you can make such changes. I'm actually little busy these days so please go ahead.

I prepared another code to avoid palette swapping but I need more time to fix all issues with this.

**Comment by idshibanov** [comment]
> Hi @idshibanov , yes you can make such changes. I'm actually little busy these days so please go ahead.
> 
> I prepared another code to avoid palette swapping but I need more time to fix all issues with this.

No worries. Cycling works and looks good now, but I did not change the update method. It still swaps the palette, so if you have ideas how to avoid it - go ahead.

**Comment by ihhub** [comment]
Hi @idshibanov , I optimized the code so we do not set palette as many times as many tiles we need to draw.

**Pull request merged**